### ISSUE TITLE
Initial work

### ISSUE THREAD

**Issue by theoweiss** [openhab/openhab-linuxpkg#2]
I've opened a PR for the initial checkin #1. I want to use this issue to discussing the further development.  Here is a to-do list with the next things which have to be done. For sure the list is not complete, just add further to-dos if they are necessary.

# Todo list for openhab b5
 * test debian packages from gradle build
   * fresh installation
   * upgrades from current cloudbees snapshot
   * upgrades from gradle snapshots
   * for openhab2-offline and openhab2-online builds
 * fix bugs in gradle build
 * remove deb and apt-repo build from openhab-distro project
 * add release task for b5
 * add release task for snapshots
 * release snapshots to bintray (replacement for apt-repo plugin)
 * integrate snapshot build and release into cloudbees
 * update projects Readme.md
 * update docs
 * test test test debian packages

# Todo list for openhab 2.0.0
 * stabilize deb build
 * build and release rpm packages

**Comment by BClark09** [comment]
Fantastic, you've been busy! Are the changes in the PR ready for an initial test? Am I right in thinking ./gradlew download-openhab2-online-snapshot && ./gradlew distro-openhab2-online-snapshot is the correct order to get a .deb file? 

> update docs

From a user's point of view, what will be different to how installation usually works which requires the documentation to change?

**Comment by theoweiss** [comment]
I've just updated the README: `./gradlew buildSnapshot` is your friend ;-).

> From a user's point of view, what will be different to how installation usually works which requires the documentation to change?

The snapshot installation will probably change because I'll try to publish the snapshots using the bintray apt-repo.

**Comment by theoweiss** [comment]
@BClark09 I've merged the initial checkin. It would be great if you could do some testing.

**Comment by BClark09** [comment]
Sure thing! I've got a spare Debian laptop ready for testing, but will get round to doing it as soon as I can.

**Comment by ThomDietrich** [comment]
Hey @theoweiss these points are of concern to me:

> - release snapshots to bintray (replacement for apt-repo plugin)
> - integrate snapshot build and release into cloudbees
> - update docs

Will identical snapshots be available on both bintray and cloudbee? If so, how long? If I understand you correctly, the idea is to provide the latest snapshot on both to ease the transition for users?

What I take from that: I'll have to switch the repository link to bintray in both the linux article and the openHABian scripts!?

**Comment by BClark09** [comment]
Testing went well:

1. I used `./gradlew buildSnapshot` to build both the OFFLINE and ONLINE .deb packages of snapshot \#624. No apparent error in building.

2. I installed openHAB on a fresh debian laptop using the offline .deb package first, no errors were logged in dpkg, or the service daemon on first run. openHAB configured itself successfully without error.

3. I installed the NTP, Yahoo Weather, and Exec1.x bindings, added some sample things and items and used mapDB persistence to restoreOnStartup. Worked flawlessly. 

4. Forced an upgrade using the same deb package, after "upgrade" bindings reinstalled themselves and persisted items displayed the expected state.

5. Repeated Steps 1-4 with the online package.

6. Repeated Steps 1-5 with the cloudbees packages, making sure to upgrade to the packages built earlier. Once again, nothing unexpected popped up during testing.

I'm going to build \#625 tomorrow so that I can test the above with specifically different snapshot files but I don't expect any differences in behaviour so from what I can see so far:

Fresh installation works.
Cloudbees > gradle upgrade works.
gradle > gradle upgrade works.

**Comment by theoweiss** [comment]
@BClark09 you(!) have been busy, thanks a lot. Sounds very good.
We also have to check the issues for the deb packaging opened in the openhab-distro project.
I found :
https://github.com/openhab/openhab-distro/issues/328

**Comment by theoweiss** [comment]
@ThomDietrich 
> Will identical snapshots be available on both bintray and cloudbee? If so, how long? If I understand you correctly, the idea is to provide the latest snapshot on both to ease the transition for users?
What I take from that: I'll have to switch the repository link to bintray in both the linux article and the openHABian scripts!?

I hope to get the bintray deployment ready before the b5 release. If I get this working the snapshots will be available for a short time on both, bintray and cloudbees to ease the transition. Together with the b5 release we will remove the snapshot build from the openhab-distro project. Hopefully. I've read the available docs how to release from gradle to bintray but found no docs or examples how to handle debian repos. As you can see there is some uncertainty if all these things are in place until the b5 release.

**Comment by BClark09** [comment]
@theoweiss 

First major problem I'm coming up with. It doesn't look like there's any handling of configuration files even though I see they're set in gradle. For example my changes to /etc/openhab2/html/index.html are lost on an upgrade.

**Comment by BClark09** [comment]
It looks like this issue is related to: https://github.com/nebula-plugins/gradle-ospackage-plugin/issues/56

**Comment by BClark09** [comment]
Alright! So Debian looks good to go,

> We also have to check the issues for the deb packaging opened in the openhab-distro project.
> I found :
> openhab/openhab-distro#328

The build doesn't create a /var/run/openhab.pid so I'm guessing this is Ubuntu specific, I'll install Ubuntu on a VM and get back to you on this.

**Comment by ThomDietrich** [comment]
I believe the pid file is related to the systemd service file [openhab2.service](https://github.com/openhab/openhab-linuxpkg/blob/master/resources/deb/systemd/openhab2.service).
http://www.dsm.fordham.edu/cgi-bin/man-cgi.pl?topic=systemd.service -> `PIDFile` and `GuessMainPID`

**Comment by BClark09** [comment]
Oh right, fair enough. Then it'll be a mixture of that assigning start.sh as it's main program and [init.d/openhab2](https://github.com/openhab/openhab-linuxpkg/blob/master/resources/deb/etc/init.d/openhab2)

**Comment by BClark09** [comment]
Whilst looking into the above, I notice that /etc/init.d/openhab2 isn't included in the gradle build. I've got a [branch I'm working on](https://github.com/BClark09/openhab-linuxpkg/commit/114f0c4a1f3b2e4e0d333d578ba15a8e26c42f3c) to address these two issues.

**Comment by BClark09** [comment]
@theoweiss I've tested this on multiple Linux distributions now (init.d systems need PR #6 included) across multiple updated and I'm pretty confident the Initial work. What are the next steps here? A release task for bintray?

**Comment by theoweiss** [comment]
@BClark09 I've just merged your PR, thanks again.

Yes we will need *two* release tasks for bintray: one for the snapshots and one for the releases. As soon as the linux-pgk build is integrated into the cloudbees build the snapshot build must be triggered automatically each time when the openhab-distro build finished successfully.

I think we should try to use the bintray gradle plugin (https://github.com/bintray/gradle-bintray-plugin). Unfortunately I found no examples for publishing deb files. Here you will find the script that I've used for publishing openHAB2 beta releases: https://github.com/openhab/openhab-linuxpkg/blob/master/resources/bin/bintray-upload-debs.sh . This must be transcribed to gradle-bintray and for snapshot releases.

Furthermore I've started to port the apt-repo maven plugin to gradle (https://github.com/theoweiss/apt-repo). Until now we used it for publishing the snapshot debs. You will find my work here https://github.com/theoweiss/apt-repo-lib/tree/initialDev and here https://github.com/theoweiss/apt-repo-gradle. The idea is to have this available in case we can't get the gradle-bintray plugin running.

If you don't mind it would be great if you could care for the gradle-bintray task and I'll try to finish the apt-repo gradle plugin as a fallback solution?

**Comment by BClark09** [comment]
Not sure when I'll be able to look at it, but I did previously have a look and think that this will get us pretty close:

```
bintray {
  user = System.env.BINTRAY_USER
  key = System.env.BINTRAY_API_KEY

  dryRun = false
  publish = true    
  override = true

  filesSpec {
    from ('build/distributions') { include '*.deb' }
       into openHAB
  }

  pkg {
    repo = 'apt-repo2'
    name = 'openhab2'
    desc = 'Packages for openHAB 2.x'
    websiteUrl = 'http://www.openhab.org/'
    issueTrackerUrl = 'https://github.com/openhab/openhab-linuxpkg/issues'
    licenses = ['EPL-1.0']
    vcsUrl = 'https://github.com/openhab/openhab-linuxpkg'
  }
}
```

Although I'm a little unsure how it will actually handle the different deb files.

**Comment by BClark09** [comment]
Hi @theoweiss, I've edited the above, but am very unsure on if this would generate anything useful. Would you mind testing this? (Probably best into the test name instead of openhab2). 

```bash
.gradlew bintrayUpload 
```

Initial thoughts:

- the .deb packages may need to be separated into four different folders in the build/distributions folder to distinguish between versions.
- If not then each package needs a specific configuration.

**Comment by theoweiss** [comment]
I had a quick look into the sources of gradle-bintray-plugin. As far as I understand the attributes will be used to create the request body: https://github.com/bintray/gradle-bintray-plugin/blob/58a2c1813ebadc8ad3242a6ca73a4fa1471823c5/src/main/groovy/com/jfrog/bintray/gradle/BintrayUploadTask.groovy#L231 . But the documentation indicates that debian information must be specified using HTTP matrix parameters or HTTP request headers (https://bintray.com/docs/api/#_debian_upload). May be using the gradle-bintray-plugin is a dead end approach :-(

**Comment by theoweiss** [comment]
May be there are two other options:

1. gradle Exec tasks https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Exec.html and make a curl call similar to that one in the upload script https://github.com/openhab/openhab-linuxpkg/blob/master/resources/bin/bintray-upload-debs.sh . 

2. the gradle rest plugin https://github.com/noamt/rest-gradle-plugin

We would also need a task to add a new version for each upload.

**Comment by BClark09** [comment]
I'd also seen the gradle Exec task as this seems to be the one used in most deb builds. I like the idea of using the REST plugin though. I might have the chance to play with both tomorrow!

**Comment by BClark09** [comment]
@theoweiss , What do you think of https://bintray.com/benclark as a way of managing these packages. 

[My own branch](https://github.com/BClark09/openhab-linuxpkg/tree/add-bintray) now does this through gradle as part of the buildSnapshot command, or manually start an upload with upload-openhab2-online-snapshot etc.

**Comment by theoweiss** [comment]
@BClark09 I'm preparing some feedback until tomorrow.

**Comment by BClark09** [comment]
@theoweiss,  did you have any quick points of feedback? I have time to work on it this weekend.

**Comment by theoweiss** [comment]
Hi @BClark09,

sorry for the delay!
First of all I think I should explain the apt-repo setup I had in my mind.
I think we should have only one repo, which is the still existing apt-repo2 for openHAB2 (https://bintray.com/openhab/apt-repo2/openhab2).
All released packages should be available from this repo. Released packages are snapshots, betas and stable versions.
Having more than one repo until now is just a workaround for technical reasons.

* The snapshots should have deb_distribution=unstable, betas have deb_distribution=testing and releases have deb_distribution=stable.
Releases have a second deb_distribution=$version. 
* Furthermore the signing support is missing X-GPG-PASSPHRASE header (just look at bintray-upload-debs.sh).
* The debian package version must not be hardcoded (def filename="${packageName}_${version}-1_all.deb").
* deb_architecture=all is not sufficient use: deb_architecture=all,armhf,armel,mips,arm64
* I think the bintray version must be created?
* calc_metadata is missing
* use expressions like this for variables (not only System.env)

````
user = project.hasProperty('bintrayUser') ? project.property('bintrayUser') : System.getenv('BINTRAY_USER')
````

Regards,
Theo

**Comment by BClark09** [comment]
Thanks @theoweiss I've started updating it now, does https://bintray.com/benclark/apt-repo2 look a little more like what you were expecting? Or would you rather everything be placed inside an openHAB2 package?

**Comment by theoweiss** [comment]
dists should  not be null: https://bintray.com/benclark/apt-repo2/openhab2-offline#files/dists
Yes place it into the openHAB2 package and separate it by versions.

**Comment by BClark09** [comment]
Alright thanks once again @theoweiss, I think we're on the same page now, sorry for the trial-error approach:

https://bintray.com/benclark/apt-repo2/openhab2

**Comment by theoweiss** [comment]
Cool, this looks very promising.

Now we have to tweak the deb creation a little bit:

* version number 2.0.0.20161218120054-1 should be "2.0.0~20161218120054-1" ( for reference <https://bintray.com/openhab/apt-repo2/openhab2/2.0.0.b3>)
* maintainer should be "openHAB <admin@openhab.org>"
* conflicts should be "openhab2-offline" for openhab2-online package and vice versa
* section should be "misc"

>sorry for the trial-error approach

It's  a good agile approach and helped to continue the discussion.

**Comment by kaikreuzer** [comment]
> maintainer should be "openHAB admin@openhab.org"

No, this actually goes nowhere really...
Does a Debian package NEED to provide an e-mail address of a maintainer? If not, I would suggest to put a link like https://community.openhab.org/ as a reference for support.

**Comment by ThomDietrich** [comment]
Is it actually still beneficial to provide **online and offline** versions of openHAB? The differences are clear to us (complete but more used space vs. internet connection needed) but maybe not clear and probably not relevant to the end user. 

From the end users perspective **two non-default** packages are *in my opinion* confusing. How about either providing a default and the other (e.g. openhab2+openhab2-offline or similar) or going the path many other repositories/package maintainers go: Defining a meta package "openhab2" pointing to the actual package openhab2-online.

Wdyt?

**Comment by BClark09** [comment]
Alright! I've got a working version @theoweiss  asked of me: https://bintray.com/benclark/apt-repo2/openhab2 with the maintainer set as suggested by @kaikreuzer. 

The description seems to be altering itself by adding on the additional properties of the Debian build, I think this might be a bug in ospackage but I think we should leave as is for now.

This is working very well when I include the following into a sources file.

```
deb https://benclark.bintray.com/apt-repo2 unstable main
``` 

apt-get will recognise same versions, and will not attempt to reinstall unless forced. It will also happily upgrade from a cloudpath version.

@ThomDietrich I like the idea, but which package do we even label as default? I have a feeling that around the world they would be used equally.

**Comment by BClark09** [comment]
What is bintray's support for snapshots? It looks they don't have an official quota for snapshots as such but the more I look into keeping the repo clean, the more I see suggestions for using bintray for releases only.

**Comment by kaikreuzer** [comment]
Bintray does not allow snapshots. They instead go to https://oss.jfrog.org.

**Comment by BClark09** [comment]
Right, so the work in #9 will need to be changed so that gradle uploads the snapshots to the oss artifactory, and the beta releases to bintray?

**Comment by theoweiss** [comment]
#9 looks good so far. I'll merge the PR because releasing betas should work. The snapshot issue could be solved with a further PR.

**Comment by theoweiss** [comment]
@kaikreuzer could you provide us some details about openHAB on oss.jfrog.org.

**Comment by kaikreuzer** [comment]
You can find infos here: https://www.jfrog.com/confluence/display/RTF/Deploying+Snapshots+to+oss.jfrog.org

tl;dr: When Bintray packages are also published to JCenter (what we do), the Bintray credentials will automatically also work for oss.jfrog.org, so that you can use that repo in your Maven pom as a deploy repo.

**Comment by kaikreuzer** [comment]
See e.g. https://github.com/openhab/openhab-distro/blob/master/pom.xml#L40

**Comment by ThomDietrich** [comment]
@kaikreuzer Any opinion on what I said earlier?

**Comment by kaikreuzer** [comment]
@ThomDietrich I agree that for openHABian we should define which version to use.
I personally prefer the offline distro very much - installation also works offline and all the setup is much quicker and less error-prone. If size is not a matter (and 16GB SD cards is the smallest you can buy nowadays...), I would go for this.
As a side note: For the Ubuntu Core Snap, we decided to use the online distro as Snaps are anyhow remotely updated and installed.

Thinking further into the future, we might also one day merge the two versions into one: Have the most common add-ons always shipped with the distro, while having the million others installable from a remote repo.

**Comment by ThomDietrich** [comment]
@kaikreuzer From your answer I get the idea that your answer to [my other question](https://github.com/openhab/openhab-linuxpkg/issues/2#issuecomment-267841567) would be "leave as is for now"!?

**Comment by kaikreuzer** [comment]
I didn't really understand what you meant with "default" packages there, so no, I tried to avoid answering that question as I didn't understand it ;-)

**Comment by ThomDietrich** [comment]
^^ touché. I was trying to propose to promote one version as the favorable one. Currently a user can select either the online or offline version but the pros and cons are unclear to users and I feel they shouldn't have to care. Hence the proposal to provide a deb package with the name "openhab2".

**Comment by kaikreuzer** [comment]
Well, then my answer still fits - I think there is only a recommendation per packaging/installation option. As said above, for snaps I would suggest "online" (that's actually the only one we offer), while for openHABian I would go for "offline".

If you want one recommendation for debian in general, I think it should be offline as well - this is what users usually expect from a debian package: That it is self-contained and does not require downloading a lot of additional stuff in order be usable, right?

**Comment by ThomDietrich** [comment]
I'm not sure about the last part (compare Firefox or other add-ons based software) but I get your argument "recommendation per packaging/installation option". I thought with the online version mostly stable this would be a good time to think about abandoning the offline version for openHAB 2.0 GA all together (or provide it as a less favorable version) but I do not want to argue about this in length here. Just wanted to point it out as this Issue / repo would be influenced by such a decision. ;)

**Comment by theoweiss** [comment]
@kaikreuzer it seems like OJO ( https://oss.jfrog.org) isn't a valid option because https://www.jfrog.com/confluence/display/RTF/Deploying+Snapshots+to+oss.jfrog.org states
"hosting your maven-compatible build snapshots". Debian repos which are not maven-compatible don't seem to be supported :-(
Some time ago you suggested to use bintray for the nightly builds because the version string doesn't contain "SNAPSHOT" anymore: https://github.com/openhab/openhab-distro/issues/270#issuecomment-244405155
From a technical perspective this would work.

**Comment by theoweiss** [comment]
As a fallback we could also stick with cloudbees and use the gradle port of the maven apt-repo plugin. It would be still better than the current solution because we could solve the lastSuccessfulBuild/WS problem which is also mentioned here https://github.com/openhab/openhab-distro/issues/270#issuecomment-244405155

**Comment by kaikreuzer** [comment]
> it seems like OJO ( https://oss.jfrog.org) isn't a valid option

Hm, good point. I didn't think about the fact that we are not talking about Maven artifacts.

> Some time ago you suggested to use bintray for the nightly builds because the version string doesn't contain "SNAPSHOT" anymore

Let's go for this. After all, Debian packages do not know the notion of a "snapshot", so it should be fine to upload all of them to Bintray.

**Comment by theoweiss** [comment]
Yes automatically removing old versions would be right solution, but for now I think it would be OK to remove them by hand using the web ui of bintray.

**Comment by BClark09** [comment]
To achieve that I'd assume it would suffice to run a script to perform the following steps:

1. Get a list of versions currently in the repo by parsing the response of: 
 `CURL -X GET "-u${BINTRAY_USER}:${BINTRAY_KEY}" "https://api.bintray.com/packages/${BINTRAY_ORG}/${BINTRAY_REPO}/${BINTRAY_PACKAGE}/`
2. List the bottom [total number of versions] - [desired number of versions] that do not match "2.0.0~b*"
3. For each match run:
 `CURL -X DELETE "-u${BINTRAY_USER}:${BINTRAY_KEY}" "https://api.bintray.com/packages/${BINTRAY_ORG}/${BINTRAY_REPO}/${BINTRAY_PACKAGE}/${versionToDelete}`

**Comment by theoweiss** [comment]
These curl calls from gradle are a bit ugly especially the [error handling](https://github.com/openhab/openhab-linuxpkg/blob/master/build.gradle#L205) and even more if some complex logic is needed. Having these calls in an "external" bash script is a way to go, but I would prefer to have these things in the gradle build script. I want to give this [rest-gradle-plugin](https://github.com/noamt/rest-gradle-plugin) a try and if I don't succeed we could take the bash script approach. What do you think?

**Comment by ThomDietrich** [comment]
Everything seems to work flawlessly till now. Great ;) Would it be possible to release a stable dummy package (e.g. the beta5 re-branded, needs to be "newer") so I can already test switching around between branches?

**Comment by theoweiss** [comment]
For testing the best way to go is:
* create your own Bintray account
* create an empty bintray Debian repository (maybe called "apt-repo2")
* within this repo create a package (maybe called "openhab2")
* create a GPG key and upload it to Bintray
* clone openhab-linuxpkg
* adapt openhab-linuxpkg to your needs (maybe with assistance from maintainers ;-) )
* release openHAB to your private repo
  ```
  ./gradlew buildSnapshot -PbintrayOrg=<thomDietrichAccount> -PbintrayRepo=apt-repo2 -PbintrayPackage=openhab2 -PbintrayUser=<thomDietrichUser> -PbintrayKey=<thomDietrichsApiKey> -PbintrayGpg=<thomDietrichsGPGKeyPassword>
  ```
* use this repo for your tests

**Comment by ThomDietrich** [comment]
Thanks for the instructions Theo but I do not intent to get active in the repository development. I've already too much on my plate, not only related to openHAB. I thought this would be a no-brainer for you guys and it wouldn't harm anyone. But it's okay, I'll just wait for the stable to be released.

**Comment by BClark09** [comment]
>I want to give this rest-gradle-plugin a try and if I don't succeed we could take the bash script approach.  What do you think?

I would also like to have everything running from within the gradle build. I particularly like the responseHandler.

A quick example:

```
task "upload-${dist}"(type: org._10ne.gradle.rest.RestTask) {

  def fileName = "${gPackageName}_${gVersion}-${gRelease}_${OSPACKAGE_ARCH}.deb"

  httpMethod = 'PUT'
  uri = 'https://api.bintray.com/content/${BINTRAY_ORG}/${BINTRAY_REPO}/${BINTRAY_PACKAGE}/${gVersion}/pool/main/${gVersion}/${fileName}'
  username = ${BINTRAY_USER}
  password = ${BINTRAY_KEY}
  requestBody = ?
  contentType = ?
  requestContentType = groovyx.net.http.ContentType.JSON
  requestHeaders = [X-GPG-PASSPHRASE: ${BINTRAY_GPG},
                    X-Bintray-Debian-Distribution: ${gDebDist},
                    X-Bintray-Debian-Component: 'main',
                    X-Bintray-Debian-Architecture: ${BINTRAY_ARCH},
                    X-Bintray-Publish: 1]
  responseHandler = {
    assert it.message == 'success'
  }
}
```

But how to add the file to the body? I couldn't be sure.

**Comment by theoweiss** [comment]
You could try:

```
 requestBody = new File(fileName)
```

for the body. 

And:

```
contentType = BINARY
```

I googled for "groovy RESTClient PUT file" because [noamt](https://github.com/noamt/rest-gradle-plugin/blob/master/src/main/groovy/org/_10ne/gradle/rest/RestTask.groovy#L39) uses "groovyx.net.http.RESTClient". Just a wild guess.

**Comment by BClark09** [comment]
Yes, you're absolutely right! I've edited the above to contain a full example. Unfortunately, my trial expired so am not able to test what it does to Bintray.

**Comment by theoweiss** [comment]
Which trial? You don't have to pay for Bintray "Community - Open Source".

**Comment by BClark09** [comment]
Whoops, my mistake. Created an account.

I've been struggling with this for a while, this is what I have so far (including a little tidy up): https://github.com/BClark09/openhab-linuxpkg/blob/add-rest-plugin/build.gradle

The request body needs to be converted to bytes to be used in requestBody (There's an issue for support of File), Since task configuration is done at the evaluation stage in gradle, I can't for the life of me work out how to use .bytes...

**Comment by theoweiss** [comment]
IMHO using .bytes may be a bad idea because I think the whole contents of the deb file would be stored in the requestBody variable, which means in RAM. Could you point me to the issue you mentioned?

**Comment by BClark09** [comment]
Certainly - https://github.com/noamt/rest-gradle-plugin/issues/26

**Comment by BClark09** [comment]
I've come to the conclusion that this won't accept a file stream for a long time. I had a play around with the underlining HTTP and REST clients and the resulting task was:

```
task "upload-${dist}"(dependsOn: "build-${dist}") {

    def fileName = "${gPackageName}_${gVersion}-${gRelease}_${OSPACKAGE_ARCH}.deb"
    def fileToUpload = new File("build/distributions/${fileName}")

    HTTPBuilder bintray = new HTTPBuilder('https://api.bintray.com', JSON)
    bintray.parser.'application/json' = { '' }
    bintray.auth.basic "$BINTRAY_USER", "$BINTRAY_KEY"

    def requestHeaders = [
        'X-GPG-PASSPHRASE':              "${BINTRAY_GPG}",
        'X-Bintray-Debian-Distribution': "${gDebDist}",
        'X-Bintray-Debian-Component':    'main',
        'X-Bintray-Debian-Architecture': "${BINTRAY_ARCH}",
        'X-Bintray-Publish':             '1'
     ]

    doLast{
        bintray.request(PUT, JSON) {
            uri.path = "/content/${BINTRAY_ORG}/${BINTRAY_REPO}/${BINTRAY_PACKAGE}/${gVersion}/pool/main/${gVersion}/${fileName}"
            headers[HttpHeaders.AUTHORIZATION] = 'Basic ' + ("$BINTRAY_USER:$BINTRAY_KEY".toString().bytes.encodeBase64())
            requestHeaders.each { key, value ->
                headers."${key}" = "${value}"
            }
            contentType = groovyx.net.http.ContentType.JSON
            requestContentType = groovyx.net.http.ContentType.BINARY
            body = fileToUpload.bytes
            response.success = {
                println 'Done.'
            }
            response.'409' = {
                err.println "WARNING: File already uploaded, new file ignored."
            }
            response.failure = {
                err.println "Failed to upload file: $it.statusLine.reasonPhrase."
                exit it.status
            }
        }    
    }
}
```

I think i'd prefer to use the curl executable task within gradle considering the memory usage in this task and the extra faff that needs to be performed to get it to function correctly. I'd also prefer to use the existing task over a bash script, wdyt?

**Comment by theoweiss** [comment]
Your right. I think the bug is in groovyx.net.http library and we would have to use the [apache hc lib](https://hc.apache.org) to get a proper solution. For now we should stick with the current curl way. What we still need is a task for removing outdated versions, where we don't have the body problem. We could use groovyx.net.http for this or a well crafted bash script.

But on the top of the todo list is the adaption of the gradle script to release 2.0.0.RC1 see here: https://community.openhab.org/t/upcoming-releases/18503/17

**Comment by theoweiss** [comment]
@BClark09 I've started to work on the adaptions needed for releasing RC1.

**Comment by BClark09** [comment]
Right, is RC1 going in Stable or Testing? Apologies, I've not had any notification on that message chain until today so was not privy to the release dates.

I was in the process of cleaning up the build.gradle (see branch [here](https://github.com/BClark09/openhab-linuxpkg/blob/build-clean/build.gradle)), but I can merge this after RC1 and also work on a way of cleaning up the old snapshot versions. I think we should be fine using the REST plugin here.

@theoweiss Would you be able to add 'userdata/etc/branding-ssh.properties' to the list of files that we dis-include from the conffile list? This came up as a problem for some people using apt.

**Comment by kaikreuzer** [comment]
> Right, is RC1 going in Stable or Testing?

I'd say Testing, just like the beta releases.

**Comment by theoweiss** [comment]
@BClark09 I'm just making the RC1 release. I've merged the necessary changes to this project. I'll do a test installation. The distribution name is "testing". If you have time tomorrow could you also make some basic tests?

**Comment by theoweiss** [comment]
> @theoweiss Would you be able to add 'userdata/etc/branding-ssh.properties' to the list of files that we dis-include from the conffile list? This came up as a problem for some people using apt.

@BClark09 now I've not been notified about your comment. We could add it to the next release or even  increment deb "release" number and make another deb release for RC1. 

I'm sure you were not aware of this community thread (https://community.openhab.org/t/upcoming-releases/18503/17m) because you haven't been in the maintainer list at the time this thread was opened. 

With a next step we have to add a parameterised task for releasing  2.x.x., shouldn't be to hard to implement.

**Comment by BClark09** [comment]
That's fine, upgrade on a debian 8 machine is working fine. Also tested a fresh install on Ubuntu 14. The packages are doing their job fine and addons seems to be working fine also.

@kaikreuzer,  after the install of legacy, 1.x bindings show up on paper-ui, but when trying to install them they stall and on refresh they no longer show even on a fresh install, can you and @theoweiss can confirm this behaviour?

> I'm sure you were not aware of this community thread (https://community.openhab.org/t/upcoming-releases/18503/17m) because you haven't been in the maintainer list at the time this thread was opened.

Initially it read "Oops! That page doesn’t exist or is private." but I can read this on refresh. Thanks.

**Comment by kaikreuzer** [comment]
> can you and @theoweiss can confirm this behaviour?

I tried the "normal" install and all seems to work well:

```
08:02:06.310 [INFO ] [smarthome.event.ExtensionEvent      ] - Extension 'binding-astro1' has been installed.
08:02:20.295 [INFO ] [b.core.service.AbstractActiveService] - Exec Refresh Service has been started
08:02:20.298 [INFO ] [smarthome.event.ExtensionEvent      ] - Extension 'binding-exec1' has been installed.
```

So I assume it is either some local problem on your end or some issue with the apt package.

**Comment by BClark09** [comment]
@kaikreuzer You're right, it was a local problem and works as expected now. But I am still getting some minor behaviour:

When the legacy switch is off in paper UI and *then* I add the legacy .kar file, (through apt, or normal installation) all the legacy bindings show up only once in the add-ons menu. If you try to install these, they will disappear until you actually turn on the legacy switch, and installation of these bindings works as expected.

**Comment by kaikreuzer** [comment]
Ok, I can live with that ;-)
